File: code/exp 3/CircConv.m
clc;
close all;
clear all;

N=input("Enter the Fundamental Period: ");
x1 = input("Enter first sequence: ");
x1 = [x1,zeros(1,(N-length(x1)))];
x2 = input("Enter second sequence: ");
x2 = [x2,zeros(1,zeros(N-length(x2)))];

y = zeros(N, 1);
%Convolution
for n = 0:N-1    
    for m=0:N-1
        h=circshift(x2,m);
        y(n+1)=y(n+1)+(x1(m+1)*h(n+1));
    end
end

X1=fft(x1);
X2=fft(x2);
Y1=X1.*X2;

disp("DFT of sequence 1: ");
disp(X1);

disp("DFT of sequence 2: ");
disp(X2);

disp("Circular convolution in time domain is ");
disp(y);

y1=IDFT(Y1, N);                                     
disp("Circular convolution in frequency domain is ");
disp(y1);

n=0:1:N-1;
subplot(2,1,1);
stem(n,abs(y));
title("Circular convolution in time domain");
ylabel("Magnitude");
xlabel("n");

subplot(2,1,2);
stem(n,abs(y1));
title("Circular convolution in frequency domain");
ylabel("Magnitude");
xlabel("n");

function y = IDFT(X, N)
    y = zeros(N,1);
    for n = 0:N-1
        for k = 0:N-1
            y(n+1) = y(n+1) + X(k+1)*1/N*exp(1i*2*pi*n*k/N);
        end
    end
end

File: code/circ_conv.c
/*
 * main.c
 */
#include <stdio.h>

int m, n, x[10], h[10], y[10], i, j, k, x2[10], a[10], swap[10];

int main(void) {
	
	printf("Enter length of first sequence \n");
	scanf("%d", &m);
	printf("Enter length of second sequence \n");
	scanf("%d", &n);

	printf("Enter the first sequence \n");
	for(i = 0; i < m; i++)
		scanf("%d", &x[i]);

	printf("Enter the second sequence \n");
	for(j = 0; j < n; j++)
		scanf("%d", &h[j]);

	if(m - n != 0){
		if(m > n){
			for(i = n; i < m; i++)
				h[i] = 0;
			n = m;
		}
		for(i = m; i < n; i++)
			x[i] = 0;
		m = n;
	}
	y[0] = 0;
	a[0] = h[0];

	for(j = 1; j<n; j++)
		a[j] = h[n - j];

	for(i = 0; k < n; k++)
		y[i] += x[i] *  a[i];

	for(k = 0; k < n; k++){
		y[k] = 0;
		for(j = 1; j < n; j++)
			x2[j] = a[j - 1];
		x2[0] = a[n - 1];

		for(i = 0; i < n; i++){
			a[i] = x2[i];
			y[k] += x[i] * x2[i];

		}


	}
//	for(i = 0; i < n; i++){
//		if(i == 0){
//			swap[i] = y[n - 1];
//			continue;
//		}
//		swap[i] = y[i - 1];
	//}

	printf("Circular convolution of x(n) and h(n) is : \n");
		for(i = 0; i < n; i++){
//
			printf("y[%d] = \t %d \n", i, y[i]);
		}

	return 0;
}
